File: myproject-server/src/myproject_server/schemas/productivity.py
from datetime import UTC, date, datetime
from typing import Annotated

from myproject_core.productivity.models import JournalType, Status
from pydantic import AfterValidator, BaseModel, ConfigDict, PlainSerializer

# --- Utilities ---


def ensure_utc(dt: datetime | None) -> datetime | None:
    """Validator to ensure that any incoming datetime is converted to UTC
    and remains timezone-aware.
    """
    if dt is None:
        return None
    if dt.tzinfo is None:
        # If the frontend forgot to send a timezone, we assume UTC
        return dt.replace(tzinfo=UTC)
    return dt.astimezone(UTC)


def serialize_utc(dt: datetime) -> str:
    """Output: Ensures outgoing datetimes have the 'Z' suffix and 3ms decimals."""
    if dt is None:
        return None
    # Ensure it's treated as UTC even if the DB returned it as naive
    if dt.tzinfo is None:
        dt = dt.replace(tzinfo=UTC)
    # Format: 2024-10-25T14:30:00.000Z
    return dt.astimezone(UTC).strftime("%Y-%m-%dT%H:%M:%S.%f")[:-3] + "Z"


# Type alias for cleaner code
UtcDatetime = Annotated[
    datetime, AfterValidator(ensure_utc), PlainSerializer(serialize_utc, when_used="json"),
]


class ProductivityBase(BaseModel):
    model_config = ConfigDict(from_attributes=True)


# --- Task Schemas ---
class TaskCreate(BaseModel):
    title: str
    description: str | None = None
    # Absolute Point in Time (UTC)
    hard_deadline: UtcDatetime | None = None

    # The Intent/Planning Day (Floating Date)
    assigned_date: date | None = None

    # The Appointment (Calendar Block - Absolute UTC)
    scheduled_start: UtcDatetime | None = None
    duration_minutes: int | None = None

    status: Status = Status.TODO
    project_ids: list[int] | None = None


class TaskUpdate(BaseModel):
    title: str | None = None
    description: str | None = None
    hard_deadline: UtcDatetime | None = None
    assigned_date: date | None = None
    scheduled_start: UtcDatetime | None = None
    duration_minutes: int | None = None
    status: Status | None = None
    completed_at: UtcDatetime | None = None


class TaskRead(ProductivityBase):
    id: int
    title: str
    description: str | None
    hard_deadline: UtcDatetime | None
    assigned_date: date | None
    scheduled_start: UtcDatetime | None
    duration_minutes: int | None
    status: Status
    created_at: UtcDatetime
    completed_at: UtcDatetime | None
    project_ids: list[int] = []


class TaskBulkUpdate(BaseModel):
    ids: list[int]
    updates: TaskUpdate
    # Optional: If provided, these projects will be ADDED to all selected tasks
    add_project_ids: list[int] | None = None
    # Optional: If provided, these projects will be REMOVED from all selected tasks
    remove_project_ids: list[int] | None = None
    # Optional: If provided, replaces all existing project links with these ones
    set_project_ids: list[int] | None = None


# --- Project Schemas ---
class ProjectCreate(BaseModel):
    name: str
    description: str | None = None
    start_date: date | None = None
    deadline: date | None = None
    status: Status = Status.TODO


class ProjectUpdate(BaseModel):
    name: str | None = None
    description: str | None = None
    start_date: date | None = None
    deadline: date | None = None
    status: Status | None = None


class ProjectRead(ProjectCreate, ProductivityBase):
    id: int


# --- Journal Schemas ---
class JournalEntryCreate(BaseModel):
    entry_type: JournalType
    reference_date: date
    title: str | None = None
    content: str
    project_id: int | None = None


class JournalEntryUpdate(BaseModel):
    title: str | None = None
    content: str | None = None
    reference_date: date | None = None
    project_id: int | None = None


class JournalEntryRead(JournalEntryCreate, ProductivityBase):
    id: int
    created_at: UtcDatetime
    updated_at: UtcDatetime

File: myproject-frontend/types/productivity.ts
export type Status = 'backlog' | 'todo' | 'in_progress' | 'completed' | 'canceled';
// Define the "Heat" order (Higher number = "Hotter" / Top of list)
export const STATUS_WEIGHTS: Record<Status, number> = {
  in_progress: 5,
  todo: 4,
  backlog: 3,
  completed: 2,
  canceled: 1,
};

export type JournalType = 'daily' | 'weekly' | 'monthly' | 'yearly' | 'project' | 'general';

export interface Project {
  id: number;
  name: string;
  description?: string;
  start_date?: string; // Floating Date: "YYYY-MM-DD"
  deadline?: string;   // Floating Date: "YYYY-MM-DD"
  status: Status;
}

export interface Task {
  id: number;
  title: string;
  description?: string;

  // Absolute Point in Time: "2024-10-25T14:30:00Z"
  // null = explicitly clear, undefined = not provided
  hard_deadline?: string | null;

  // Floating Planning Date: "YYYY-MM-DD"
  // null = explicitly clear, undefined = not provided
  assigned_date?: string | null;

  // Absolute Appointment Time: "2024-10-25T09:00:00Z"
  // null = explicitly clear, undefined = not provided
  scheduled_start?: string | null;

  duration_minutes?: number;
  status: Status;

  // Metadata: Always ISO Strings
  created_at: string;
  completed_at?: string;

  project_ids: number[];
}

export interface JournalEntry {
  id: number;
  entry_type: JournalType;
  reference_date: string; // Floating Date: "YYYY-MM-DD"
  title?: string;
  content: string;
  project_id?: number;
  created_at: string;
  updated_at: string;
}
